service/client.go

```diff
@@ -160,28 +183,40 @@ func (c *Client) Connect() error {
 	}
 
 	c.wsClient = wsClient
-	c.receiveCh = make(chan ClientMessage, msgChSize)
-	c.errorCh = make(chan error)
 
-	var wg sync.WaitGroup
-	wg.Add(1)
+	c.wg.Add(2)
+
 	go func() {
-		defer wg.Done()
-		c.msgReader()
+		defer c.wg.Done()
+		c.msgReader(wsClient)
+		c.mut.Lock()
+		if c.wsClient != nil {
+			c.wsClient = nil
+			c.mut.Unlock()
+			c.reconnectHandler()
+			return
+		}
+		c.mut.Unlock()
 	}()
 
 	go func() {
-		for err := range c.wsClient.ErrorCh() {
+		defer c.wg.Done()
+		for err := range wsClient.ErrorCh() {
 			c.sendError(err)
 		}
-		wg.Wait()
-		close(c.errorCh)
 	}()
 
 	return nil
 }
 
 func (c *Client) Send(msg ClientMessage) error {
+	c.mut.RLock()
+	defer c.mut.RUnlock()
```

Comment:
Is there a reason these are added now?
Or, I think I'm not understanding what the `c.mut` is specifically locking... Is it the `c.closed` bool?

Reply:
Well, the reason is that now we are touching state from multiple goroutines and need to be extra careful. Of course the logic inevitably becomes more complex.

In fact, the reconnect handler runs in the reader goroutine and could trigger concurrently with any other operation (e.g. Close).

Once you start allowing for concurrency then there's no turning back really so I had to also make sure every method was safe to be called concurrently with eachother

Reply:
Ok, gotcha -- I missed that we were touching from multiple goroutines. Now it makes sense (here and all the other places).